src/Features/LanguageServer/Protocol/Handler/SignatureHelp/SignatureHelpHandler.cs

```diff
@@ -53,48 +53,73 @@ internal class SignatureHelpHandler : ILspServiceDocumentRequestHandler<LSP.Text
             var triggerInfo = new SignatureHelpTriggerInfo(SignatureHelpTriggerReason.InvokeSignatureHelpCommand);
             var options = _globalOptions.GetSignatureHelpOptions(document.Project.Language);
 
+            var sigHelpItems = new ArrayBuilder<SignatureHelpItems>();
+
             foreach (var provider in providers)
             {
                 var items = await provider.Value.GetItemsAsync(document, position, triggerInfo, options, cancellationToken).ConfigureAwait(false);
 
-                if (items != null)
+                if (items is null)
                 {
-                    var sigInfos = new ArrayBuilder<LSP.SignatureInformation>();
+                    continue;
+                }
 
-                    foreach (var item in items.Items)
-                    {
-                        LSP.SignatureInformation sigInfo;
-                        if (clientCapabilities.HasVisualStudioLspCapability() == true)
-                        {
-                            sigInfo = new LSP.VSInternalSignatureInformation
-                            {
-                                ColorizedLabel = GetSignatureClassifiedText(item)
-                            };
-                        }
-                        else
-                        {
-                            sigInfo = new LSP.SignatureInformation();
-                        }
+                sigHelpItems.Add(items);
+            }
+
+            // Multiple providers could be offering items at the current location. As
+            // there might be an ObjectCreationExpression within a MethodInvocation.
+            // We want to identify the items nearest to the cursor.
+
+            SignatureHelpItems? helpItems = null;
+            var minDistance = int.MaxValue;
+            foreach (var items in sigHelpItems.ToArrayAndFree())
+            {
+                var distanceFromCursor = position - items.ApplicableSpan.Start;
+                if (distanceFromCursor < minDistance)
```

Comment:
So idea - the logic here isn't super complicated so I'm not sure how much I mind, but I was thinking it would be nice to have some kind of overall `ISignatureHelpService` that is responsible for aggregating the providers and choosing the best ones.  We could then have unified logic for choosing that and share it with the code in the signature help controller too - https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/Core.Wpf/SignatureHelp/Controller.Session_ComputeModel.cs#L29

Or if not a whole service thing - maybe some shared helpers for iterating the providers and choosing.

Reply:
Added a service that took inspiration from the QuickInfoService. The services is integrated into the LSP handler and the EditorFeatures controller. Existing tests should cover it.